src/providers/gce/mod.rs

```diff
@@ -14,82 +14,122 @@
-        .set_hostname_if_exists(hostname)
-        .add_ssh_keys(ssh_keys)?
-        .build())
-}
+        Ok(GceProvider { client })
+    }
 
-fn fetch_all_ssh_keys(client: &retry::Client) -> Result<Vec<String>> {
-    let keys = fetch_ssh_keys(client, "instance/attributes/sshKeys")?;
-    if !keys.is_empty() {
-        return Ok(keys);
+    fn endpoint_for(name: &str) -> String {
+        format!("http://metadata.google.internal/computeMetadata/v1/{}", name)
     }
-    let mut keys = fetch_ssh_keys(client, "instance/attributes/ssh-keys")?;
 
-    let block_project_keys: Option<String> = client.clone().get(retry::Raw, url_for_key("instance/attributes/block-project-ssh-keys")).send()?;
+    fn fetch_all_ssh_keys(&self) -> Result<Vec<String>> {
+        let keys = self.fetch_ssh_keys("instance/attributes/sshKeys")?;
+        if !keys.is_empty() {
+            return Ok(keys);
+        }
+        let mut keys = self.fetch_ssh_keys("instance/attributes/ssh-keys")?;
 
-    if block_project_keys == Some("true".to_owned()) {
-        return Ok(keys);
-    }
+        let block_project_keys: Option<String> = self.client
+            .clone()
+            .get(retry::Raw, GceProvider::endpoint_for("instance/attributes/block-project-ssh-keys"))
+            .send()?;
 
-    keys.append(&mut fetch_ssh_keys(client, "project/attributes/sshKeys")?);
+        if block_project_keys == Some("true".to_owned()) {
+            return Ok(keys);
+        }
+
+        keys.append(&mut self.fetch_ssh_keys("project/attributes/sshKeys")?);
+
+        Ok(keys)
+    }
 
-    Ok(keys)
+    fn fetch_ssh_keys(&self, key: &str) -> Result<Vec<String>> {
+        let key_data: Option<String> = self.client.get(retry::Raw, GceProvider::endpoint_for(key)).send()?;
+        if let Some(key_data) = key_data {
+            let mut keys = Vec::new();
+            for l in key_data.lines() {
+                if l.is_empty() {
+                    continue
+                }
+                let mut l = l.to_owned();
+                let index = l.find(':')
+                    .ok_or("character ':' not found in line in key data")?;
+                keys.push(l.split_off(index+1));
+            }
+            Ok(keys)
+        } else {
+            // The user must have not provided any keys
+            Ok(Vec::new())
+        }
+    }
 }
 
-fn fetch_ssh_keys(client: &retry::Client, key: &str) -> Result<Vec<String>> {
-    let key_data: Option<String> = client.get(retry::Raw, url_for_key(key)).send()?;
-    if let Some(key_data) = key_data {
-        let mut keys = Vec::new();
-        for l in key_data.lines() {
-            if l.is_empty() {
-                continue
+impl MetadataProvider for GceProvider {
+    fn attributes(&self) -> Result<HashMap<String, String>> {
+        let mut out = HashMap::with_capacity(3);
+
+        let add_value = |map: &mut HashMap<_, _>, key: &str, name| -> Result<()> {
```

Comment:
Unfortunate that there is so much duplication with the `add_value` closure, but I'm not sure if there is a better way to do it after thinking through some of the alternatives.